### PR TITLE
Logging via logging module

### PR DESCRIPTION
**Pull request by pbertera**
This patch adds support for python [logging](https://docs.python.org/2/library/logging.html) module.
This makes PyPXE compatibile with software using the logging module.
This patch is backward compatible: you're not obliged to use logging in your project: each module can instantiate a local (StreamHandler](https://docs.python.org/2/library/logging.handlers.html#streamhandler) if no logger is received in the constructor.

--syslog and --syslog-port system options are added.

### CODE REVIEW

**Comment by psychomario** [comment]
This all looks good to me.

**Comment by psychomario** [comment]
Does the change from
```options = message.split(chr(0))[2: -1]```
to
```options = message.split(chr(0))[3: -1]```
in `tftp.py` fix a bug I'm unaware of?

**Comment by psychomario** [comment]
Does the change from
```options = message.split(chr(0))[2: -1]```
to
```options = message.split(chr(0))[3: -1]```
in `tftp.py` fix a bug I'm unaware of?

**Comment by psychomario** [comment]
Does the change from
```options = message.split(chr(0))[2: -1]```
to
```options = message.split(chr(0))[3: -1]```
in `tftp.py` fix a bug I'm unaware of?

**Comment by psychomario** [comment]
Does the change from
```options = message.split(chr(0))[2: -1]```
to
```options = message.split(chr(0))[3: -1]```
in `tftp.py` fix a bug I'm unaware of?

**Comment by psychomario** [comment]
(Apologies for the repeated message)

**Comment by psychomario** [comment]
Looks like some odd merge artifact. I'll ignore it for now, and check it in testing.

File: pypxe-server.py [hunk 1]
```diff
@@ -85,9 +89,22 @@
             dargs.update(loadedcfg)
             args = argparse.Namespace(**dargs)
 
+        # setup logger
+        logger = logging.getLogger(sys.argv[0])
```
Comment (mmattioli) [inline pypxe-server.py]: Why are you using `sys.argv[0]` here? Should you be using `argparse` like everywhere else? I could be misunderstanding this line but please clarify. Thanks.

File: pypxe/dhcp.py [hunk 2]
```diff
@@ -37,30 +37,43 @@ def __init__(self, **serverSettings):
...
-            print '\tDHCP Subnet Mask: {}'.format(self.subnetmask)
-            print '\tDHCP Router: {}'.format(self.router)
-            print '\tDHCP DNS Server: {}'.format(self.dnsserver)
-            print '\tDHCP Broadcast Address: {}'.format(self.broadcast)
-            print '\tDHCP File Server IP: {}'.format(self.fileserver)
-            print '\tDHCP File Name: {}'.format(self.filename)
-            print '\tProxyDHCP Mode: {}'.format(self.mode_proxy)
-            print '\tUsing iPXE: {}'.format(self.ipxe)
-            print '\tUsing HTTP Server: {}'.format(self.http)
+        self.logger.debug('NOTICE: DHCP server started in debug mode. DHCP server is using the following:')
```
Comment (mmattioli) [inline pypxe/dhcp.py]: In L64-L76 should we remove the `\t` characters since we're using the logger now?

File: pypxe/tftp.py [hunk 3]
```diff
@@ -18,18 +18,31 @@ class TFTPD:
     '''
     def __init__(self, **serverSettings):
         self.ip = serverSettings.get('ip', '0.0.0.0')
-        self.port = serverSettings.get('port', 69)
+        self.port = serverSettings.get('port', 79)
```
Comment (mmattioli) [inline pypxe/tftp.py]: Why did you change the port to 79?

File: pypxe/dhcp.py [hunk 4]
```diff
@@ -37,30 +37,43 @@ def __init__(self, **serverSettings):
...
+        self.logger.debug('\tDHCP Lease Range: ' + self.offerfrom + ' - ' + self.offerto)
+        self.logger.debug('\tDHCP Subnet Mask: ' + self.subnetmask)
+        self.logger.debug('\tDHCP Router: ' + self.router)
+        self.logger.debug('\tDHCP DNS Server: ' + self.dnsserver)
+        self.logger.debug('\tDHCP Broadcast Address: ' + self.broadcast)
+        self.logger.debug('\tDHCP File Server IP: ' + self.fileserver)
+        self.logger.debug('\tDHCP File Name: ' + self.filename)
+        self.logger.debug('\tProxyDHCP Mode: ' + str(self.mode_proxy))
+        self.logger.debug('\tUsing iPXE: ' + str(self.ipxe))
+        self.logger.debug('\tUsing HTTP Server: ' + str(self.http))
```
Comment (mmattioli) [inline pypxe/dhcp.py]: In L65-L76 should we remove the `\t` characters since we're using the logger now? If using the logger perhaps we shouldn't use the `\t`? Thoughts?

File: pypxe/dhcp.py [hunk 5]
```diff
@@ -218,11 +232,11 @@ def dhcpOffer(self, message):
...
-        if self.mode_debug:
-            print '[DEBUG] DHCPOFFER - Sending the following'
-            print '\t<--BEGIN HEADER-->\n\t{headerResponse}\n\t<--END HEADER-->'.format(headerResponse = repr(headerResponse))
-            print '\t<--BEGIN OPTIONS-->\n\t{optionsResponse}\n\t<--END OPTIONS-->'.format(optionsResponse = repr(optionsResponse))
-            print '\t<--BEGIN RESPONSE-->\n\t{response}\n\t<--END RESPONSE-->'.format(response = repr(response))
+        
+        self.logger.debug('DHCPOFFER - Sending the following')
+        self.logger.debug('\t<--BEGIN HEADER-->\n\t{headerResponse}\n\t<--END HEADER-->'.format(headerResponse = repr(headerResponse)))
+        self.logger.debug('\t<--BEGIN OPTIONS-->\n\t{optionsResponse}\n\t<--END OPTIONS-->'.format(optionsResponse = repr(optionsResponse)))
+        self.logger.debug('\t<--BEGIN RESPONSE-->\n\t{response}\n\t<--END RESPONSE-->'.format(response = repr(response)))
```
Comment (mmattioli) [inline pypxe/dhcp.py]: In L236-L239 should we remove the `\t` characters since we're using the logger now? If using the logger perhaps we shouldn't use the `\t`? Thoughts?

File: pypxe/dhcp.py [hunk 6]
```diff
@@ -230,11 +244,11 @@ def dhcpAck(self, message):
...
-        if self.mode_debug:
-            print '[DEBUG] DHCPACK - Sending the following'
-            print '\t<--BEGIN HEADER-->\n\t{headerResponse}\n\t<--END HEADER-->'.format(headerResponse = repr(headerResponse))
-            print '\t<--BEGIN OPTIONS-->\n\t{optionsResponse}\n\t<--END OPTIONS-->'.format(optionsResponse = repr(optionsResponse))
-            print '\t<--BEGIN RESPONSE-->\n\t{response}\n\t<--END RESPONSE-->'.format(response = repr(response))
+        
+        self.logger.debug('DHCPACK - Sending the following')
+        self.logger.debug('\t<--BEGIN HEADER-->\n\t{headerResponse}\n\t<--END HEADER-->'.format(headerResponse = repr(headerResponse)))
+        self.logger.debug('\t<--BEGIN OPTIONS-->\n\t{optionsResponse}\n\t<--END OPTIONS-->'.format(optionsResponse = repr(optionsResponse)))
+        self.logger.debug('\t<--BEGIN RESPONSE-->\n\t{response}\n\t<--END RESPONSE-->'.format(response = repr(response)))
```
Comment (mmattioli) [inline pypxe/dhcp.py]: In L249-L251 should we remove the `\t` characters since we're using the logger now? If using the logger perhaps we shouldn't use the `\t`? Thoughts?

File: pypxe/http.py [hunk 7]
```diff
@@ -29,18 +30,29 @@ def __init__(self, **serverSettings):
...
+            handler.setFormatter(formatter)
+            self.logger.addHandler(handler)
+
+            if self.mode_debug:
+                self.logger.setLevel(logging.DEBUG)
+
+        self.logger.debug('NOTICE: HTTP server started in debug mode. HTTP server is using the following:')
+        self.logger.debug('\tHTTP Server IP: {}'.format(self.ip))
+        self.logger.debug('\tHTTP Server Port: {}'.format(self.port))
+        self.logger.debug('\tHTTP Network Boot Directory: {}'.format(self.netbootDirectory))
```
Comment (mmattioli) [inline pypxe/http.py]: In L47-L49 should we remove the `\t` characters since we're using the logger now? If using the logger perhaps we shouldn't use the `\t`? Thoughts?

File: pypxe/tftp.py [hunk 8]
```diff
@@ -18,18 +18,31 @@ class TFTPD:
...
+            handler.setFormatter(formatter)
+            self.logger.addHandler(handler)
+
+            if self.mode_debug:
+                self.logger.setLevel(logging.DEBUG)
+
+        self.logger.debug('TFTP server started in debug mode. TFTP server is using the following:')
+        self.logger.debug('\tTFTP Server IP: {}'.format(self.ip))
+        self.logger.debug('\tTFTP Server Port: {}'.format(self.port))
+        self.logger.debug('\tTFTP Network Boot Directory: {}'.format(self.netbootDirectory))
```
Comment (mmattioli) [inline pypxe/tftp.py]: In L43-L45 should we remove the `\t` characters since we're using the logger now? If using the logger perhaps we shouldn't use the `\t`? Thoughts?

File: pypxe/dhcp.py (hunk 4 above)
Comment (psychomario) [inline pypxe/dhcp.py]: I'm not sure about this one, it does improve readability. Here's some test output:
```
2015-04-15 20:44:06,145 INFO Starting TFTP server...
2015-04-15 20:44:06,145 DEBUG TFTP server started in debug mode. TFTP server is using the following:
2015-04-15 20:44:06,145 DEBUG   TFTP Server IP: 0.0.0.0
2015-04-15 20:44:06,145 DEBUG   TFTP Server Port: 79
2015-04-15 20:44:06,146 DEBUG   TFTP Network Boot Directory: .
2015-04-15 19:44:06,146 INFO Starting DHCP server...
2015-04-15 19:44:06,146 DEBUG NOTICE: DHCP server started in debug mode. DHCP server is using the following:
2015-04-15 19:44:06,146 DEBUG   DHCP Server IP: 192.168.2.2
2015-04-15 19:44:06,146 DEBUG   DHCP Server Port: 67
2015-04-15 19:44:06,146 DEBUG   DHCP Lease Range: 192.168.2.100 - 192.168.2.150
2015-04-15 19:44:06,146 DEBUG   DHCP Subnet Mask: 255.255.255.0
2015-04-15 19:44:06,146 DEBUG   DHCP Router: 192.168.2.1
2015-04-15 19:44:06,146 DEBUG   DHCP DNS Server: 8.8.8.8
2015-04-15 19:44:06,146 DEBUG   DHCP Broadcast Address: <broadcast>
2015-04-15 19:44:06,146 DEBUG   DHCP File Server IP: 192.168.2.2
2015-04-15 19:44:06,147 DEBUG   DHCP File Name: pxelinux.0
2015-04-15 19:44:06,147 DEBUG   ProxyDHCP Mode: False
2015-04-15 19:44:06,147 DEBUG   Using iPXE: False
2015-04-15 19:44:06,147 DEBUG   Using HTTP Server: False
2015-04-15 19:44:06,147 INFO PyPXE successfully initialized and running!
```

However, I'm not against removing this.

**Comment by mmattioli** [comment]
@psychomario @pbertera I highlighted a few areas where the `'t` was preserved and sent to the logger but didn't comment on every single area where it appeared. Thoughts on preserving that? I'm thinking it should be kept for when referring back to the logs since it will visually make them easier to read but also considering removing them since the logger might mess up the output? Now that I'm typing this out perhaps we should just test and see which works better...

**Comment by psychomario** [comment]
There is a bug present in the debug lines. I'm trying to locate it now. The granularity added in #54 seems to be missing. Is this a problem with  the merge from `master` to `development`?

Example:

```sudo python2 pypxe-server.py --debug http --dhcp
2015-04-15 20:44:06,145 INFO Starting TFTP server...
2015-04-15 20:44:06,145 DEBUG TFTP server started in debug mode. TFTP server is using the following:
2015-04-15 20:44:06,145 DEBUG   TFTP Server IP: 0.0.0.0
2015-04-15 20:44:06,145 DEBUG   TFTP Server Port: 79
2015-04-15 20:44:06,146 DEBUG   TFTP Network Boot Directory: .
2015-04-15 19:44:06,146 INFO Starting DHCP server...
2015-04-15 19:44:06,146 DEBUG NOTICE: DHCP server started in debug mode. DHCP server is using the following:
2015-04-15 19:44:06,146 DEBUG   DHCP Server IP: 192.168.2.2
2015-04-15 19:44:06,146 DEBUG   DHCP Server Port: 67
2015-04-15 19:44:06,146 DEBUG   DHCP Lease Range: 192.168.2.100 - 192.168.2.150
2015-04-15 19:44:06,146 DEBUG   DHCP Subnet Mask: 255.255.255.0
2015-04-15 19:44:06,146 DEBUG   DHCP Router: 192.168.2.1
2015-04-15 19:44:06,146 DEBUG   DHCP DNS Server: 8.8.8.8
2015-04-15 19:44:06,146 DEBUG   DHCP Broadcast Address: <broadcast>
2015-04-15 19:44:06,146 DEBUG   DHCP File Server IP: 192.168.2.2
2015-04-15 19:44:06,147 DEBUG   DHCP File Name: pxelinux.0
2015-04-15 19:44:06,147 DEBUG   ProxyDHCP Mode: False
2015-04-15 19:44:06,147 DEBUG   Using iPXE: False
2015-04-15 19:44:06,147 DEBUG   Using HTTP Server: False
2015-04-15 19:44:06,147 INFO PyPXE successfully initialized and running
```
We should not be seeing the DEBUG for tftp

**Comment by psychomario** [comment]
@mmattioli I commented on one of the code blocks in reply to your comment. [here](https://github.com/psychomario/PyPXE/pull/61#discussion_r28455374). Also my comment above this one shows what it looks like. This is for stdout though, not sure how syslog will look

**Comment by mmattioli** [comment]
@psychomario I just saw, let's leave the `\t` for now since it makes the logs easier to read. I'm looking at the bug you just pointed out.

File: pypxe-server.py (hunk 1 above)
Comment (pbertera) [inline pypxe-server.py]: This the logger name sys.argv[0] is the command name.. maybe we can use a fixed string like "PyPXE".

File: pypxe/tftp.py (hunk 3 above)
Comment (pbertera) [inline pypxe/tftp.py]: Yes, a mistake: should not commited, I have another TFTP running on my machine..

File: pypxe/tftp.py (hunk 8 above)
Comment (mmattioli) [inline pypxe/tftp.py]: @psychomario look at lines 42-45, `self.logger.debug` I think that's the issue. Those lines should have the if `self.mode_debug` statement preceding them.

**Comment by psychomario** [comment]
Regarding the bug, It looks like to me that if we have a valid self.logger, we don't enforce granularity. And L102-3 of `pypxe-server.py` uses DEBUG if there is any value set. I propose removing L102-3, and unindenting the `if self.mode_debug` by one level in each {dhcp, tftp, http}.py

File: pypxe-server.py (hunk 1 above)
Comment (mmattioli) [inline pypxe-server.py]: If running as such from the command line `$ python pypxe-server.py` then wouldn't `sys.argv[0]` simply be the string "pypxe-server.py"? If so then that's fine but we should confirm that if this is run twice that the second time it doesn't overwrite the original log. Can we either make a new log or add onto the previous log?

**Comment by pbertera** [comment]
Humm. Yes, my patch breaks the granularity log. I found the issue.

**Comment by mmattioli** [comment]
@pbertera @psychomario is [what I found here](https://github.com/psychomario/PyPXE/pull/61/files#r28456309) the issue? I just want to make sure we're talking about the same thing because I don't se L102-3 of `pypxe-server.py` to be a problem? I could be missing or misinterpreting something so please clarify. Thanks.

**Comment by pbertera** [comment]
about the "\t" in log messages: doesn't causes big issues using a StreamHandler.
But can cause encoding issues using a syslog handler: Syslog messages are displayed like:
````
Apr 16 01:32:31 2015-04-15 21: 27:30,048 DEBUG NOTICE: TFTP server started in debug mode. TFTP server is using the following:
Apr 16 01:32:31 2015-04-15 21: 27:30,048 DEBUG #011TFTP Server IP: 0.0.0.0
```
Where #011 is the \t char. IMHO not a big issue

**Comment by mmattioli** [comment]
I think we should either replace the `\t` with two whitespace characters `  `, two `--` or just remove it. The `#011` is technically a bug and shouldn't be there.

**Comment by psychomario** [comment]
Apparently the line numbers on my local copy are screwed up. I'll hardlink to @pbertera's dev branch. So [L102-3](https://github.com/pbertera/PyPXE/blob/development/pypxe-server.py#L102-L103) sets the debug level (`.debug()` doesn't output if level isn't `logging.DEBUG`), but only `if args.MODE_DEBUG`, which is now a csv string rather than a bool.

When we get to [L39-40](https://github.com/pbertera/PyPXE/blob/development/pypxe/tftp.py#L39-L40) in `tftp.py` (applies to all of the submodules), the logging is already at `DEBUG` level, so these lines have no effect for any logging type. We never actually get to this condition anyway because for any debug string we create a `self.logger`, thereby falsifying the parent `if`.

**Comment by psychomario** [comment]
I vote for two whitespace over `--`. `--` Isn't as clean for seperating the indent levels visually.

**Comment by pbertera** [comment]
Yes, two whitespace can be a good solution.

**Comment by mmattioli** [comment]
@psychomario I understand now. If you turn the logger on, it checks it _any_ of the debug statements were passed in and automatically sets the mode of the logger to debug even if only one of the services are passed into the debug argument. Once it's set to debug it's passed into all of the services when their threads are started thereby making all of the services log at the debug level even if only one service was passed in to be on the debug level.

Is there a way around this? I can't seem to think of or find a clean way to keep everything within one instance of the logger while maintaining granularity for debugging.

**Comment by psychomario** [comment]
See [this](https://github.com/psychomario/PyPXE/pull/61#issuecomment-93550003) comment. We remove the `setLevel` if from `pypxe-server` (we don't have to, we could check for `all`, but that's done elsewhere already), and unindent the conditional `setLevel` in each of the submodules.

**Comment by pbertera** [comment]
@mmattioli seems right: we need different logger instances for each service if we wants granularity..

**Comment by mmattioli** [comment]
@psychomario if we are passing in _one_ instance of the logger to each of the services then once _any_ of them sets the mode to debug it will set it for all of the services because they all use _one_ logger instance.

Part of me feels like abandoning the use of the python logger because there's no clean way to preserve granularity and if users wants to log output they can just pipe all output to a file as such

```bash
$ python pypxe-server.py --debug http --dhcp > output.txt
```

**Comment by psychomario** [comment]
Apologies, I was wrong, I didnt realise there was only one instance.

**Comment by pbertera** [comment]
@mmattioli yes in a "standalone" usage there isn't too many advantages using the logging module.
The main advantage comes when one need to integrate PyPXE in bigger projects using the logging module.

**Comment by pbertera** [comment]
Tomorrow I'll work on it to find a more clever solution.

**Comment by psychomario** [comment]
The logging module allows for logging namespaces (`getChild`) which can have different levels.
See the following snippet:

```
import logging
root = logging.getLogger()
handler = logging.StreamHandler()
formatter = logging.Formatter('%(asctime)s %(name)s %(levelname)s %(message)s')
handler.setFormatter(formatter)
root.addHandler(handler)

root.setLevel(logging.INFO)
tftp = root.getChild("tftp").setLevel(logging.DEBUG)

root.debug("ROOT")
root.getChild("tftp").debug("TFTP")
```

The addition of `%(name)s` to the format string allows more specific debug messages without manually adding them to the `.debug()/.info()` string.

**Comment by mmattioli** [comment]
@pbertera, while we strive to have PyPXE both a standalone solution and available as a library so that it can be used in other projects it should fall on the overarching project to handle logging. Think about it, if you wrote something and want to keep track of logging for your application then why would you want to have to check individual logs for each of the libraries you're using?

When used standalone, I think it's best to just pipe the output to a file; when used as a library I think it should fall on the overarching project to handle its own logging as PyPXE will provide the log info, there just needs to be something written to handle it if it's even desired to be captured or handled.

**Comment by mmattioli** [comment]
@pbertera, following @psychomario's proposal, would you want to rebase and recommit using namespaces? If so I'll leave this PR open so you can update it.

**Comment by psychomario** [comment]
@mmattioli While I agree with you, this PR will allow syslog logging, therefore centralizing logging.

**Comment by mmattioli** [comment]
@psychomario I'm not opposed to using the logger if we can find a clean solution; what you've found regarding namespaces looks like a solution which is why I suggested he rebase and recommit (if he rebases his development branch off of our development branch, effectively starting over, and recommits then those commits will appear in this PR which is why I said I'd leave it open)?

**Comment by pbertera** [comment]
@mmattioli @psychomario yes, seems a clean solution. tomorrow I'll work on it. I'll came back with a new commit. Thanks,

**Comment by mmattioli** [comment]
@pbertera sounds good :+1:

File: pypxe/tftp.py [hunk 9]
```diff
@@ -140,7 +148,6 @@ def listen(self):
             #Resent those that have timed out
             for i in self.ongoing:
                 if self.ongoing[i]['timeout']+5 < time.time() and self.ongoing[i]['retries']:
-                    print self.ongoing[i]['handle'].tell()
```
Comment (pbertera) [inline pypxe/tftp.py]: I removed this line. looks a debug print, or maybe has some meaning ?

Comment (psychomario) [inline pypxe/tftp.py]: This line can be safely removed, it was left in when I rewrote this section. Good catch.

**Comment by mmattioli** [comment]
@pbertera I assume you closed because you're going to make a new branch off of `psychomario:development` since there was a recent pull and then open a new PR with that? If so, that's cool just be sure to reference this PR in your new one.
